src/responder.ts

```diff
@@ -1,60 +1,297 @@
+     * @param appointmentId The id of the Appointment this object is responding to.
+     * @param ethereumResponse The IEthereumResponse containing what needs to be submitted.
+     * @param [confirmationsRequired=40] The number of confirmations required before a transaction is trusted.
+     * @param [maxAttempts=10] The maximum number of retries before the Responder will give up.
+     */
+    constructor(
+        readonly signer: ethers.Signer,
+        public readonly appointmentId: string,
+        public readonly ethereumResponse: IEthereumResponse,
+
+        public readonly confirmationsRequired = 40,
+        private readonly maxAttempts: number = 10
+    ) {
+        super(signer, appointmentId, ethereumResponse);
+
+        signer.provider.on("block", this.updateLastBlockTime);
+    }
+
+    /**
+     * This method should be called when an instance is disposed.
+     */
+    public destroy() {
+        this.signer.provider.removeListener("block", this.updateLastBlockTime);
+    }
+
+    private updateLastBlockTime() {
+        this.timeOfLastBlock = Date.now();
+    }
+
+    protected async respond() {
+        while (this.mAttemptsDone < this.maxAttempts) {
+            this.mAttemptsDone++;
+            try {
+                // Try to call submitStateFunction, but timeout with an error if
+                // there is no response for 30 seconds.
+                const tx = await promiseTimeout(this.submitStateFunction(), 30000);
+
+                // The response has been sent, but should not be considered confirmed yet.
+                this.responseFlow.status = ResponseState.ResponseSent;
+                this.asyncEmit("responseSent", this.responseFlow);
+
+                // Wait for enough confirmations before declaring success
+                await new Promise((resolve, reject) => {
+                    const provider = this.signer.provider;
+
+                    const cleanup = () => {
+                        provider.removeListener("block", newBlockHandler);
+                        clearInterval(intervalHandle);
+                    }
+
+                    const newBlockHandler = async () => {
+                        const receipt = await provider.getTransactionReceipt(tx.hash);
+                        if (receipt == null) {
+                            // There was a re-org, consider this attempt failed and attempt the transaction again
+                            cleanup();
+                            reject(new ReorgError(tx, "There could have been a re-org, the transaction was sent but was later not found."));
+                        } else if (receipt.confirmations >= this.confirmationsRequired) {
+                            cleanup();
+                            resolve();
+                        }
+                    };
+                    provider.on("block", newBlockHandler);
+
+                    const intervalHandle = setInterval( () => {
```

Comment:
Also, at the moment, given the current parameters, if no block is received for 60 secs an error will be thrown and the attempt failed. However after 1 sec we will start the next loop again, at which point this condition will be checked again, and probably fail straight away, we're basically giving the system one extra second to notice a block being mined when it has already failed for 60 seconds. Is this the intention?

Reply:
Perhaps not the cleanest way, but after all, if there are no new blocks, this is probably going to need the upper management to do something, so retrying many times and failing fast might give time to address it elsewhere (e.g.: spawn another responder); it might make sense to abort straight away after long enough time without blocks.

Reply:
KK, ideally we should be aborting in a more explicit way, this will do for now though.